src/circuit/gadget/ecc/chip/mul/incomplete.rs

```diff
@@ -0,0 +1,419 @@
+use std::ops::Deref;
+
+use super::super::{copy, CellValue, EccConfig, EccPoint, Var};
+use super::{INCOMPLETE_HI_RANGE, INCOMPLETE_LO_RANGE, X, Y, Z};
+use ff::Field;
+use halo2::{
+    circuit::Region,
+    plonk::{
+        Advice, Column, ConstraintSystem, Error, Expression, Fixed, Permutation, VirtualCells,
+    },
+    poly::Rotation,
+};
+
+use pasta_curves::{arithmetic::FieldExt, pallas};
+
+pub(super) struct Config {
+    // Number of bits covered by this incomplete range.
+    num_bits: usize,
+    // Selector used to constrain the cells used in incomplete addition.
+    pub(super) q_mul: Column<Fixed>,
+    // Cumulative sum used to decompose the scalar.
+    pub(super) z: Column<Advice>,
+    // x-coordinate of the accumulator in each double-and-add iteration.
+    pub(super) x_a: Column<Advice>,
+    // x-coordinate of the point being added in each double-and-add iteration.
+    pub(super) x_p: Column<Advice>,
+    // y-coordinate of the point being added in each double-and-add iteration.
+    pub(super) y_p: Column<Advice>,
+    // lambda1 in each double-and-add iteration.
+    pub(super) lambda1: Column<Advice>,
+    // lambda2 in each double-and-add iteration.
+    pub(super) lambda2: Column<Advice>,
+    // Permutation
+    pub(super) perm: Permutation,
+}
+
+// Columns used in processing the `hi` bits of the scalar.
+// `x_p, y_p` are shared across the `hi` and `lo` halves.
+pub(super) struct HiConfig(Config);
+impl From<&EccConfig> for HiConfig {
+    fn from(ecc_config: &EccConfig) -> Self {
+        let config = Config {
+            num_bits: INCOMPLETE_HI_RANGE.len(),
+            q_mul: ecc_config.q_mul_hi,
+            x_p: ecc_config.advices[0],
+            y_p: ecc_config.advices[1],
+            z: ecc_config.advices[9],
+            x_a: ecc_config.advices[3],
+            lambda1: ecc_config.advices[4],
+            lambda2: ecc_config.advices[5],
```

Comment:
What's the rationale for this ordering?

Reply:
The main rationale is that "hi_config z cannot overlap with complete addition columns", which is checked in an assertion in `mul.rs` (L86). Were it not for this requirement, we could have used a more natural ordering (e.g. [2], [3], [4], [5]).

Reply:
The rationale is to avoid an overlap with the output columns of complete addition, which are used to initialise the accumulator at 2[T]. This has now been explained at https://github.com/zcash/orchard/pull/111/commits/b696163e318d541f1327d851e2e9cb8f1b33287a.